### PR TITLE
ACSL stackpointer support

### PR DESCRIPTION
**Pull request by woosh**
This PR closes issue #12. It introduces stackpointer support for ACSL contract generation.

The main idea is to parse the source, and then transform the AST before the program is translated into horn clauses for the solver. The contract inferred for the transformed source is then translated back into a contract for the original source.

### Things To Notice

Bumped version to 0.4.0

### Implementation
The transformation consist of creating a new function acting on global variables instead of arguments that are pointers to the stack. This function is referred to as the *transformed function*. In addition to the transformed function a so called *wrapper function* is also created. The wrapper function copies values from the stack to the global variables, invokes the transformed function, and then copies the values from the global variables back to the stack. Any callsite of the original function containing stack pointer arguments will be replaced by a call to the wrapper function.

The major implementation changes are as follows. 

#### AST transformations
`CCAstStackPtrArgToGlobalTransformer`: This is the driving class for the AST transformation. It replaces function calls containing stack pointers with calls to the propper wrapped function. For this to work properly, identifiers in the AST needs to be annotated with type information. In order to accomodate this, a special internal rule `EvarWithType` is introduced in `concurrentC.cf`.

`CCAstTypeAnnotator`: This class annotates the AST with type information. It replaces any identifier with a node containing the identifier and its type information.

Besides these two classes there is a set of utility classes that perform various AST operations.

#### Back-translation
Book-keeping of changes made to the AST in order to be able to translate inferred contracts back to the original source is largely realized by two classes:

`CallsiteTransform`: Each instance contains all the knowledge that is needed to transform a specific function at a specific call site in order to replace arguments pointing towards the stack (stack pointer arguments), with global variables.
  
`AstAdditions`: AstAdditions contains all the additions that needs to be added to the AST in order for it to contain all functions and variables used by the program after stack pointers have been replaced. This includes, among other things, introduced global variables, functions acting upon them, which parameters that were replaced etc. Instances of `AstAdditions` are created by instances of `CallsiteTransforms`.

`MergeTransformedFunctionsContracts`: This class is the main transformation of inferred contracts back to the original source. Since one function can potentially give rise to many transformed functions, in case of several arguments, and also since the original function may also still be invoked by with only heap arguments, we need to combine all these contracts somehow. This class takes care of that by performing a "meet" between all the contracts involved.

#### Tests
A number of tests are added in the `regeression-tests/horn-contracts` folder. There are also some edge cases added to `toh-contract-translation` that were discovered during development.

### Limitations

1. The solution does not work for recursive functions. This is because the transformed function will call the wrapper function with a pointer argument that points to a global variable.

2. If the original function returns some pointer it got as an argument, and that pointer is a stack pointer, this return value will be broken. In such a case the original function would return the pointer to the stack, but the transformed one will return the address of the global variable replacing the stack pointer.

3. The current implementation does not work if the function is invoked through a pointer. In such a case we don't know the name of the function being invoked and cannot add a transformed function with global variables.

### Discussed goals
Prior to implementation a stretch goal was discussed,

1. Output contracts in ACSL AST, and implement preprocessors &    back-translators using this.

This goal has **not** been achieved.

### CODE REVIEW

**Review by zafer-esen — COMMENTED** [review]
Thank you for this significant contribution, it looks great overall. I have one minor comment, after addressing that it is good to go from my side.

File: regression-tests/horn-hcc-2/fischer.hcc [hunk 1]
```diff
@@ -1,4 +1,5 @@
 int lock = 0;
+int tid = _;
```
[the remark this answers is not in the document]
Comment (woosh) [reply regression-tests/horn-hcc-2/fischer.hcc:2]: Right. This `thread` extension is unfamiliar to me, but I made the change because I get an "Undeclared identifier" during the type annotation without it. 
```
tricera.concurrency.TypeAnnotationException: Undeclared identifier in expression: tid, line: 10 col:28
```
Syntactically `thread` only needs a `CIdent` between the `[]`. However, that identifier is used in integer context on line 14. My idea was that adding the definition of `tid` with a nondeterministic initialization would keep the semantics, but that was just an assumption on my part.

[the remark this answers is not in the document]
Comment (woosh) [reply regression-tests/horn-hcc-2/fischer.hcc:2]: The problem occurs with this PR because `tid` is used as an integer variable on line 14. When the type annotator is parsing that (and the type annotator is introduced with this PR) it can't find any integer variable `tid` because it has never been declared/defined. I'll change the type annotator to take care of that. Question: Should the type of the thread ID be an `int`, an `unsigned int`, `size_t` or some other type?

[the remark this answers is not in the document]
Comment (zafer-esen) [reply regression-tests/horn-hcc-2/fischer.hcc:2]: Thanks, I think the thread ID can always be a mathematical integer, maybe this needs to be handled in `CCReader` after parsing.

[the remark this answers is not in the document]
Comment (woosh) [reply regression-tests/horn-hcc-2/fischer.hcc:2]: I had a look in `CCReader`. As far as I understand, the thread id is mapped to a `CCVar` of type `CCInt`. (See https://github.com/uuverifiers/tricera/blob/8f518e8e262466cf9904ac9747193c702a830155/src/tricera/concurrency/CCReader.scala#L1124, the only place where the CHC translator accesses this part of the AST.) Based on that my suggestion is to let the type annotator map it to an ordinary `int`. Would you agree?

[the remark this answers is not in the document]
Comment (woosh) [reply regression-tests/horn-hcc-2/fischer.hcc]: I pushed a new version where the tread id is mapped to int and where the `tid` definition is removed from `fischer.hcc`.

[the remark this answers is not in the document]
Comment (woosh) [reply regression-tests/horn-hcc-2/fischer.hcc]: Yes!
